### ISSUE TITLE
SPA route behaviour + bindata for static files

### ISSUE THREAD

**Issue by speedwheel** [kataras/iris#1068]
I am trying to find the fastest way to serve my static files when doing a SPA.
All the tests are done using the starter template from Vue Cli 3 (when you create a new project) and bombardier with the following command:
`bombardier -n 100000 -c 125 http://localhost:8080`

Here is what I have tried so far:

```go
assetHandler := b.StaticHandler("./web/public", false, false)
app.SPA(assetHandler)
```
1. This works just fine, but slower speed compared to the other method. 13k req/s.

```go
var (
    homeBindata []byte
)

func init() {
    homeBindata, _ = bootstrap.GzipAsset("web/public/index.html")
}

// I tried and app.StaticWeb("/", "./web/public")
app.StaticEmbeddedGzip("/", "./web/public", GzipAsset, GzipAssetNames)
app.Get("/", function(ctx context.Context) {
    ctx.ContentType("text/html")
    ctx.ResponseWriter().Header().Add("Vary", "Accept-Encoding")
    ctx.ResponseWriter().Header().Add("Content-Encoding", "gzip")
    ctx.Write(homeBindata)
})
```
2. This works almost fine, the only disadvantage is that if you are on different page, for example "/about", you get a 404 error, But I get 50k req/s, which is nice.

```go
app.Get("/{f:path}", function(ctx context.Context) {
    ctx.ContentType("text/html")
    ctx.ResponseWriter().Header().Add("Vary", "Accept-Encoding")
    ctx.ResponseWriter().Header().Add("Content-Encoding", "gzip")
    ctx.Write(data)
})
```
3. This loads the index.html at 50k/req, but doesn't loads the assets.

```go
var (
    homeBindata []byte
)

func init() {
    homeBindata, _ = bootstrap.GzipAsset("web/public/index.html")
}

app.Get("/{f:path}", function(ctx context.Context) {
    assetFolder := "web/public"
    requestPath := ctx.Path()
    filePath := assetFolder + requestPath
    if requestPath == "/" {
        WriteIndex(ctx)
     return
    }
    _, err := bootstrap.GzipAsset(filePath)
    if err != nil {
    WriteIndex(ctx)
    return
    }
    ctx.ServeFile(filePath, true)
})

func WriteIndex(ctx iris.Context) {
   ctx.ContentType("text/html")
    ctx.ResponseWriter().Header().Add("Vary", "Accept-Encoding")
    ctx.ResponseWriter().Header().Add("Content-Encoding", "gzip")
    ctx.Write(homeBindata)
}

```
4. This method almost works perfectly, I get the 50k req/s, I can reload the page and not get a 404, it loads the rest of the static files, but it's not using bindata (`ctx.ServeFile()` is slower), so I am getting very close to the perfect combination. I know that I could serve the rest of the files also using ` ctx.Write()`, but I would have to write the headers manually for each individual type of file which is not such a big deal, but I was wondering if there is any other easier way.

So basically I want to combine the benefits from the first and second method so I am able to server all the static files (html, css, jss, img etc with bindata), but in the same time to have the "/{f:path}" benefits, meaning that if I go to any url that is not a css, js, img etc. it should default to the index.html.

**Comment by kataras** [comment]
Nice to see that @speedwheel , we use almost the same pattern for the iris website itself!

OK, look how you can do it as fast as possible:

Secondly, you define a simple method which will send the raw bytes from kataras/bindata or the original go-bindata:
```go
func htmlBindata(b []byte) iris.Handler {
	return func(ctx iris.Context) {
		context.AddGzipHeaders(ctx.ResponseWriter()) // import "github.com/kataras/iris/context"
		ctx.ContentType(context.ContentHTMLHeaderValue)
		ctx.Write(b)
	}
}
```

and you define the pages in a simple map with handlers:

```go
	// Register pages and their handlers.
	var (
		pages = map[string]iris.Handler{
			"/favicon.ico": htmlBindata(_bindataWebPublicImagesFaviconico),
			"/":            htmlBindata(_bindataWebPublicIndexhtml),
			"index.html":   htmlBindata(_bindataWebPublicIndexhtml),
		}
	)
```



You need to handle the `404` and the `assets` which are bindata as well, this is how you define a handler that we will use inside our main handler when not known page is requested:

```go
assetHandler := iris.StaticEmbeddedHandler("./web/public", Asset, AssetNames, false)
```

And finally your main handler:
```go
	app.Any("/{resource:path}", func(ctx iris.Context) {
		if h, ok := pages[ctx.Path()]; ok {
			h(ctx)
			return
		}

		// 3 days for assets (259200 = 60s x 60m x 24h x 3d)
		// ctx.Header("cache-control", "public, max-age=259200")
		assetHandler(ctx) // this handles not found too
	})

```

That's all.

**Comment by speedwheel** [comment]
Thanks for the elegant solution, I've just tested it now, but when I got to the "/about" route for example, which is handled by vue, and refresh the page, I get a 404 error, so I made some little changes (this is because I don't want to update the map everytime to point to the index.html when I add a new route inside vue, especially if they will be dynamic):

```go
app.Any("/{resource:path}", func(ctx iris.Context) {
		var isAsset bool
		if h, ok := pages[ctx.Path()]; ok {
			h(ctx)
			return
		}
              
              // check if it's an asset
		for _, asset := GzipAssetNames() {
			if ctx.Path() == strings.Replace(asset, "web/public", "", 1) {
				isAsset = true
				break
			}
		}
		if !isAsset {
			pages["/"](ctx)
			return
		}
		// 3 days for assets (259200 = 60s x 60m x 24h x 3d)
		// ctx.Header("cache-control", "public, max-age=259200")
		AssetHandler(ctx) // this handles not found too
	})
```
